src/main/java/org/broadinstitute/hellbender/utils/locusiterator/SamplePartitioner.java

```diff
@@ -0,0 +1,142 @@
+package org.broadinstitute.hellbender.utils.locusiterator;
+
+import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.downsampling.Downsampler;
+import org.broadinstitute.hellbender.utils.downsampling.PassThroughDownsampler;
+import org.broadinstitute.hellbender.utils.downsampling.ReservoirDownsampler;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+
+import java.util.*;
+
+/**
+ * Divides reads by sample and (if requested) does a preliminary downsampling pass
+ * with a ReservoirDownsampler.
+ *
+ * Note: stores reads by sample ID string, not by sample object
+ */
+final class SamplePartitioner<T extends GATKRead> {
```

Comment:
This class does not need to be parameterized any more now that we have a unified read interface (the type parameter is only to handle both `SAMRecord` and `GATKSAMRecord` in GATK3) -- can you refactor out the type parameter?